flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io;
+
+import org.apache.flink.annotation.Internal;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
+ *
+ * <p>For the most efficient usage, user of this class is suppose to call {@link #pollNext()}
+ * until it returns that no more elements are available. If that happens, he should check if
+ * input {@link #isFinished()}. If not, he should wait for {@link #isAvailable()}
+ * {@link CompletableFuture} to be completed. For example:
+ *
+ * <pre>
+ * {@code
+ *	while (!input.isFinished()) {
+ *		AsyncDataInput<Optional<T>> input = ...;
+ *		Optional<T> next;
+ *
+ *		while (true) {
+ *			next = input.pollNext();
+ *			if (!next.isPresent()) {
+ *				break;
+ *			}
+ *			// do something with next
+ *		}
+ *
+ *		input.isAvailable().get();
+ *	}
+ * }
+ * </pre>
+ */
+@Internal
+public interface AsyncDataInput<T> {
+	CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
+
+	/**
+	 * Poll the next element.
+	 *
+	 * @return
+	 * <ol>
+	 *     <li>
+	 *         If the {@link T} is an {@code Optional}: return {@code Optional.of(element)} -
+	 *         next returned element. {@code Optional.empty()} will be returned if there is no data
+	 *         to return or if {@link #isFinished()} returns true.
+	 *     </li>
+	 *     <li>
+	 *         If next element or {@code null} if there is no data to return or if
+	 *         {@link #isFinished()} returns true.
+	 *     </li>
+	 * </ol>
+	 */
+	T pollNext() throws Exception;
```

Comment:
Instead of using `AsyncDataInput<Optional<T>>`, would it make sense to make this method explicitly return `Optional<T>`? This way, the contract in your java doc would be "enforced" by method signature explicitly.

Reply:
The `Input` interface in #8476 PR uses `AsyncDataInput<StreamElement>` and uses `null` as sentinel (the interface doesn't use `Optional<StreamElement>` due to performance concerns).

We have discussed offline with @pnowojski and there were some alternative ideas how the interface of `AsyncDataInput` may be more explicit (so the contract  of `pollNext()` method is not overloaded by `Optional.empty()` or `null` return value as marker) and more cleaner (at least, from my point of view).

I'll just leave one of the ideas that I find fine (but not enforced).
The `AsyncDataInput<T>` could have been extended to have two `poll*` methods, with the same behavior regarding traversal of input, but with just different caller conventions. So it's up to the caller to decide which flavor they prefer: 
```java
interface AsyncDataInput<T> {
    ...
    Optional<T> pollNext();
    
    @Nullable
    default T pollNextOrNull() {
         return pollNext().orElse(null);
    }
}
```

Reply:
I have split this into `AsyncDataInput` and `NullableAsyncDataInput`.